### ISSUE TITLE
Does not support TPT Inheritance

### ISSUE THREAD

**Issue by TefoT4** [jcachat/EntityFramework.DynamicFilters#32]
Before i get to my problem, i want to say thank you for the grate work you are doing on this project. We really appreciate all the effort. :)

The problem i'm experiencing is that i have implemented the Table Per Type Inheritance and i have a simplified model below.

public interface IEntity
{
        bool IsDeleted { get; set; }
}

public class Person : IEntity
{
        public int id { get; set; }
        public bool IsDeleted { get; set; }
}

public class Instructor : Person
{
        public DateTime HireDate { get; set; }
}

public class Student : Person
{
        public DateTime EnrollmentDate { get; set; }
}

public class FilterDbContext : DbContext
{
        public FilterDbContext()
        {
           this.EnableAllFilters();
        }

        public DbSet<Person> People { get; set; }
        public DbSet<Instructor> Instructors { get; set; }
        public DbSet<Student> Students { get; set; }

        protected override void OnModelCreating(DbModelBuilder modelBuilder)
        {
            modelBuilder.Entity<Person>().ToTable("People");
            modelBuilder.Entity<Instructor>().ToTable("Instructors");
            modelBuilder.Entity<Student>().ToTable("Students");

           modelBuilder.Filter("IsDeleted", (IEntity e) => e.IsDeleted, false);
        }
}

private static void Main(string[] args)
{
     var context = new FilterDbContext();
     var instructors = context.Instructors.ToList();
}

![model](https://cloud.githubusercontent.com/assets/9367920/7589638/236f6776-f8c4-11e4-8f32-a8de3e829d49.PNG)

I want to get all instructors where isDeleted = false. but as you can see, the isDeleted property is on the Person Model and the Instructor model inherits from Person. However on the Database, person is its on table and Instructor is also its on table with its primary key being a foreign key to the person table.

**Comment by jcachat** [comment]
What happens when you call "context.Instructors.ToList()"?  Is it not filtering on IsDeleted or throwing an error?

**Comment by jcachat** [comment]
I set up a test using the models and filter exactly as you posted here.  Everything is working correctly for me.  The filter is applied when I query from Students, Instructors, and People.  I'm gong to need some more information from you about what exactly is the problem you are seeing as you didn't state what the problem is, just that you are having one...  :)

**Comment by TefoT4** [comment]
I added two instructor records to the Db then marked the last record as deleted.
It threw and exception. Property "IsDeleted" not found on EdmType Instructor.

**Comment by jcachat** [comment]
I just pushed up the test I created to try to duplicate this.  Can you take a look at that and compare it to what you are doing?  We must be doing something different since it works for me...  Or if you can prepare a full solution to demonstrate it and email it to me (jcachat12@gmail.com), that would work too.

**Comment by TefoT4** [comment]
Let me prep up a solution. I will email it soon as i'm done. :)

**Comment by saltstorm** [comment]
I am also having the same problem, did you ever find a solution to this?

**Comment by jcachat** [comment]
No, I am not able to reproduce the problem with the models TefoT4 posted and still waiting on a full solution that demonstrates the problem. If you have more information on how to reproduce it, please post it and I'll take a look.

**Comment by saltstorm** [comment]
My setup is *slightly* different but should work the same. I get the error found in [LambdaToDbExpressionVisitor line 248](https://github.com/jcachat/EntityFramework.DynamicFilters/blob/b3ef786f6a8b11e6a46ed236d5a906c1cd0bdbe9/src/EntityFramework.DynamicFilters/LambdaToDbExpressionVisitor.cs#L248) formatted like this: Property Deleted not found in Entity Type BaseModel

Now, that is strange to me because I do not see my setup as accessing entities outside of the main model?
```
    protected override void OnModelCreating(DbModelBuilder modelBuilder)
    {
        modelBuilder.Filter("SoftDelete", (BaseModel x) => x.Deleted == null);
        base.OnModelCreating(modelBuilder);
    }
```

```
    public class BaseModel
    {
        public DateTime? Created { get; set; }

        [MaxLength(32)]
        public String CreatedBy { get; set; }

        public DateTime? Updated { get; set; }

        [MaxLength(32)]
        public String UpdatedBy { get; set; }

        [Index]
        public DateTime? Deleted { get; set; }
    }
```

```
    public class Person : BaseModel
    {
        public int Id { get; set; }

        [Required, Display(Name = "First Name"), MaxLength(32)]
        public string FirstName { get; set; }

        [Display(Name = "Middle Name"), MaxLength(64)]
        public string MiddleName { get; set; }

        [Required, Display(Name = "Last Name"), MaxLength(32)]
        public string LastName { get; set
    }

    [Table("Residents")]
    public class Resident : Person
    {
        [Required, MaxLength(4)]
        public string Title { get; set; }

        [MaxLength(32), Display(Name = "Maiden Name")]
        public string MaidenName { get; set; }

        [MaxLength(32), Display(Name = "Famailiar Name")]
        public string FamiliarName { get; set; }

        [Required, MaxLength(1)]
        public string Gender { get; set; } // M or F

        [Display(Name = "Date of Birth"), DataType(DataType.Date)]
        public DateTime? DateOfBirth { get; set; }
    }
```

I wrote this up pretty quickly before work, I will happily spend some time to help you reproduce the error I get if this doesn't help later today.

Thanks, 

Tom.

**Comment by tomsharratt** [comment]
After further investigation the problem seems to be that the 'Resident' entity doesn't have the properties that 'BaseModel' has because it is inheriting them from 'Person'. 'Person' does have the properties from 'BaseModel'. So when the code tries to find the 'Deleted' property on 'Resident' it just falls over and dies.

This still seems strange because you can obviously access the 'Deleted' property in the linq queries.

**Comment by jcachat** [comment]
I tried reproducing this using the models and filter exactly as you posted and it works fine for me...  What database and which version of EF are you using (I tried with 6.1.2 and 6.1.3)?  I may need you to package up a full solution and email it to me (jcachat12@gmail.com) so I can figure out what's different.

**Comment by tomsharratt** [comment]
Ef 6.1.3 & sql server

**Comment by tomsharratt** [comment]
Could you send me what you did to reproduce (tsharratt55@gmail.com)?

**Comment by jcachat** [comment]
The problem seems to be caused by you having a DbSet<Person> created in your DbContext in addition to the DbSet<Resident>.  If you remove that, it works fine.  I'm not sure yet why that causes a problem but I'll dig into it and figure it out.  It might be a couple days though - I'm actually on vacation right now and traveling tomorrow.

**Comment by tomsharratt** [comment]
"DbSet created in your DbContext in addition to the DbSet" - Sorry, what do you mean? Don't I have to have my DbSet's defined in my DbContext?

**Comment by jcachat** [comment]
In your ApplicationDbContext class, you have these 2 lines:
```
public DbSet<Person> People { get; set; }
public DbSet<Resident> Residents { get; set; }
```

At least in the project you sent me, "People" is not referenced so you can safely remove it.  But, that obviously changes your table structure and will result in a single Residents table instead of 2 tables which is probably not what you're going for.

Just a theory/guess, but... I think the use of TPT between People and Residents and the inheritance (non-TPT) between People and BaseModel is what is causing the issue.  But no idea if it's something inside EF or with my filters.  If you want to work around this (for now at least), you could move everything from BaseModel into People, create an IBaseModel interface for those fields, and then create the filter against that interface.

**Comment by tomsharratt** [comment]
That is exactly what I thought too... i'll give that work around you suggest for now, i'll let you know how it goes.

**Comment by tomsharratt** [comment]
I am still getting the same problem when I create IBaseModel and add the implementation to Person.

**Comment by jcachat** [comment]
Found the problem.  Fix is pushed and NuGet has been updated.

The problem was being caused by TPT inheritance.  Basically, when you query out "Residents", EF is automatically including a join against Person.  Since both of those C# classes inherit the deleted filter, it was being applied by each of those 2 DbScanExpressions.  And where the Residents C# class has the Deleted property, the Residents database table does not - hence the error.  So the fix was to detect this and not apply the filter where it should not be.

Another note (and this may apply more to @TefoT4):  Because of how the queries are parsed, the problem is only hit if you create a filter like this:
```
modelBuilder.Filter("IsDeleted", (IEntity e) => e.IsDeleted == false)
```
as opposed to this:
```
modelBuilder.Filter("IsDeleted", (IEntity e) => e.IsDeleted, false)
```
The difference is one uses a parameter and the other does not.  When I changed his filter to the non-parameter version, I hit the same problem with his models.  I'm wondering if the trouble reproducing that case was because of this.
